### ISSUE TITLE
java.lang.NoClassDefFoundError: org/apache/http/client/methods/HttpUriRequest since 5.1.2

### ISSUE THREAD

**Issue by sipi** [owlcs/owlapi#744]
Hi,

I have the following errors (at execution time) since version 5.1.2 :

```
dataRangeComplementOf(fr.lirmm.graphik.graal.io.owl.test.OWL2ParserTest)  Time elapsed: 0.148 sec  <<< ERROR!
java.lang.NoClassDefFoundError: org/apache/http/client/methods/HttpUriRequest
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.eclipse.rdf4j.rio.jsonld.JSONLDParser.parse(JSONLDParser.java:104)
	at org.semanticweb.owlapi.rio.RioParserImpl.parseDocumentSource(RioParserImpl.java:171)
	at org.semanticweb.owlapi.rio.RioParserImpl.parse(RioParserImpl.java:125)
	at uk.ac.manchester.cs.owl.owlapi.OWLOntologyFactoryImpl.loadOWLOntology(OWLOntologyFactoryImpl.java:192)
	at uk.ac.manchester.cs.owl.owlapi.OWLOntologyManagerImpl.load(OWLOntologyManagerImpl.java:1069)
	at uk.ac.manchester.cs.owl.owlapi.OWLOntologyManagerImpl.loadOntology(OWLOntologyManagerImpl.java:1031)
	at uk.ac.manchester.cs.owl.owlapi.OWLOntologyManagerImpl.loadOntologyFromOntologyDocument(OWLOntologyManagerImpl.java:981)
	at uk.ac.manchester.cs.owl.owlapi.OWLOntologyManagerImpl.loadOntologyFromOntologyDocument(OWLOntologyManagerImpl.java:1001)
	at fr.lirmm.graphik.graal.io.owl.OWL2Parser.<init>(OWL2Parser.java:128)
	at fr.lirmm.graphik.graal.io.owl.OWL2Parser.<init>(OWL2Parser.java:169)
	at fr.lirmm.graphik.graal.io.owl.test.OWL2ParserTest.dataRangeComplementOf(OWL2ParserTest.java:1799)
```

My project tests success with the following dependency
```
              <dependency>
			<groupId>net.sourceforge.owlapi</groupId>
			<artifactId>owlapi-apibinding</artifactId>
			<version>5.1.1</version>
		</dependency>
```
But fails with 
```
              <dependency>
			<groupId>net.sourceforge.owlapi</groupId>
			<artifactId>owlapi-apibinding</artifactId>
			<version>5.1.2</version>
		</dependency>
```

There is no errors at compile time.

**Comment by ansell** [comment]
Can you run ``mvn dependency:tree`` on your project with both 5.1.1 and 5.1.2 and include the two outputs here?

**Comment by ignazio1977** [comment]
The missing class is part of httpclient and the stack trace suggests this might be related to the version change for the rdf4j version - is it possible that another dependency is interfering by forcing an incompatible httpclient version?

in 5.1.2, httpclient 4.5.3 is included.

**Comment by sipi** [comment]
Thanks for your help,

@ansell 
#### output of mvn dependency:tree for 5.1.1
```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ graal-io-owl ---
[INFO] fr.lirmm.graphik:graal-io-owl:jar:1.3.2-SNAPSHOT
[INFO] +- fr.lirmm.graphik:graal-util:jar:1.3.2-SNAPSHOT:compile
[INFO] |  \- org.jgrapht:jgrapht-core:jar:1.1.0:compile
[INFO] +- fr.lirmm.graphik:graal-core:jar:1.3.2-SNAPSHOT:compile
[INFO] +- fr.lirmm.graphik:graal-api:jar:1.3.2-SNAPSHOT:compile
[INFO] +- net.sourceforge.owlapi:owlapi-apibinding:jar:5.1.1:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-api:jar:5.1.1:compile
[INFO] |  |  +- org.tukaani:xz:jar:1.6:compile
[INFO] |  |  \- org.apache.commons:commons-rdf-api:jar:0.3.0-incubating:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-impl:jar:5.1.1:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-parsers:jar:5.1.1:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-oboformat:jar:5.1.1:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-tools:jar:5.1.1:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-rio:jar:5.1.1:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-model:jar:2.1.4:compile
[INFO] |  |  |  \- org.eclipse.rdf4j:rdf4j-util:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-api:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-languages:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-datatypes:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-binary:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-n3:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-nquads:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-ntriples:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-rdfjson:jar:2.1.4:compile
[INFO] |  |  |  \- com.fasterxml.jackson.core:jackson-core:jar:2.8.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-rdfxml:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-trix:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-turtle:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-trig:jar:2.1.4:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-jsonld:jar:2.1.4:compile
[INFO] |  |  \- com.github.jsonld-java:jsonld-java:jar:0.9.0:compile
[INFO] |  |     +- com.fasterxml.jackson.core:jackson-databind:jar:2.7.4:compile
[INFO] |  |     |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.7.0:compile
[INFO] |  |     +- org.apache.httpcomponents:httpclient-osgi:jar:4.5.2:compile
[INFO] |  |     |  +- org.apache.httpcomponents:httpclient:jar:4.5.2:compile
[INFO] |  |     |  +- commons-codec:commons-codec:jar:1.9:compile
[INFO] |  |     |  +- org.apache.httpcomponents:httpmime:jar:4.5.2:compile
[INFO] |  |     |  +- org.apache.httpcomponents:httpclient-cache:jar:4.5.2:compile
[INFO] |  |     |  \- org.apache.httpcomponents:fluent-hc:jar:4.5.2:compile
[INFO] |  |     +- org.apache.httpcomponents:httpcore-osgi:jar:4.4.4:compile
[INFO] |  |     |  +- org.apache.httpcomponents:httpcore:jar:4.4.4:compile
[INFO] |  |     |  \- org.apache.httpcomponents:httpcore-nio:jar:4.4.4:compile
[INFO] |  |     \- org.slf4j:jcl-over-slf4j:jar:1.7.21:runtime
[INFO] |  +- net.sf.trove4j:trove4j:jar:3.0.3:compile
[INFO] |  +- com.github.ben-manes.caffeine:caffeine:jar:2.3.5:compile
[INFO] |  +- com.google.guava:guava:jar:20.0:compile
[INFO] |  +- com.google.inject:guice:jar:4.1.0:compile
[INFO] |  |  +- javax.inject:javax.inject:jar:1:compile
[INFO] |  |  \- aopalliance:aopalliance:jar:1.0:compile
[INFO] |  +- com.google.inject.extensions:guice-assistedinject:jar:4.1.0:compile
[INFO] |  +- com.google.inject.extensions:guice-multibindings:jar:4.1.0:compile
[INFO] |  +- com.google.code.findbugs:jsr305:jar:3.0.1:compile
[INFO] |  \- commons-io:commons-io:jar:2.5:compile
[INFO] +- org.apache.commons:commons-lang3:jar:3.7:compile
[INFO] +- org.apache.commons:commons-collections4:jar:4.1:compile
[INFO] +- org.slf4j:slf4j-api:jar:1.7.25:compile
[INFO] \- junit:junit:jar:4.11:test
[INFO]    \- org.hamcrest:hamcrest-core:jar:1.3:test
```

#### and for 5.1.2
```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ graal-io-owl ---
[INFO] fr.lirmm.graphik:graal-io-owl:jar:1.3.2-SNAPSHOT
[INFO] +- fr.lirmm.graphik:graal-util:jar:1.3.2-SNAPSHOT:compile
[INFO] |  \- org.jgrapht:jgrapht-core:jar:1.1.0:compile
[INFO] +- fr.lirmm.graphik:graal-core:jar:1.3.2-SNAPSHOT:compile
[INFO] +- fr.lirmm.graphik:graal-api:jar:1.3.2-SNAPSHOT:compile
[INFO] +- net.sourceforge.owlapi:owlapi-apibinding:jar:5.1.2:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-api:jar:5.1.2:compile
[INFO] |  |  +- org.tukaani:xz:jar:1.6:compile
[INFO] |  |  \- org.apache.commons:commons-rdf-api:jar:0.3.0-incubating:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-impl:jar:5.1.2:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-parsers:jar:5.1.2:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-oboformat:jar:5.1.2:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-tools:jar:5.1.2:compile
[INFO] |  +- net.sourceforge.owlapi:owlapi-rio:jar:5.1.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-model:jar:2.2.2:compile
[INFO] |  |  |  \- org.eclipse.rdf4j:rdf4j-util:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-api:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-languages:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-datatypes:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-binary:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-n3:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-nquads:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-ntriples:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-rdfjson:jar:2.2.2:compile
[INFO] |  |  |  \- com.fasterxml.jackson.core:jackson-core:jar:2.8.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-rdfxml:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-trix:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-turtle:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-trig:jar:2.2.2:compile
[INFO] |  |  +- org.eclipse.rdf4j:rdf4j-rio-jsonld:jar:2.2.2:compile
[INFO] |  |  \- com.github.jsonld-java:jsonld-java:jar:0.11.1:compile
[INFO] |  |     +- com.fasterxml.jackson.core:jackson-databind:jar:2.9.0:compile
[INFO] |  |     |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.9.0:compile
[INFO] |  |     +- org.apache.httpcomponents:httpclient-osgi:jar:4.5.3:compile
[INFO] |  |     |  +- org.apache.httpcomponents:httpclient:jar:4.5.3:compile
[INFO] |  |     |  +- commons-codec:commons-codec:jar:1.9:compile
[INFO] |  |     |  +- org.apache.httpcomponents:httpmime:jar:4.5.3:compile
[INFO] |  |     |  +- org.apache.httpcomponents:httpclient-cache:jar:4.5.3:compile
[INFO] |  |     |  \- org.apache.httpcomponents:fluent-hc:jar:4.5.3:compile
[INFO] |  |     +- org.apache.httpcomponents:httpcore-osgi:jar:4.4.6:compile
[INFO] |  |     |  +- org.apache.httpcomponents:httpcore:jar:4.4.6:compile
[INFO] |  |     |  \- org.apache.httpcomponents:httpcore-nio:jar:4.4.6:compile
[INFO] |  |     \- org.slf4j:jcl-over-slf4j:jar:1.7.25:runtime
[INFO] |  +- net.sf.trove4j:trove4j:jar:3.0.3:compile
[INFO] |  +- com.github.ben-manes.caffeine:caffeine:jar:2.3.5:compile
[INFO] |  +- com.google.guava:guava:jar:20.0:compile
[INFO] |  +- com.google.inject:guice:jar:4.1.0:compile
[INFO] |  |  +- javax.inject:javax.inject:jar:1:compile
[INFO] |  |  \- aopalliance:aopalliance:jar:1.0:compile
[INFO] |  +- com.google.inject.extensions:guice-assistedinject:jar:4.1.0:compile
[INFO] |  +- com.google.inject.extensions:guice-multibindings:jar:4.1.0:compile
[INFO] |  +- com.google.code.findbugs:jsr305:jar:3.0.1:compile
[INFO] |  \- commons-io:commons-io:jar:2.5:compile
[INFO] +- org.apache.commons:commons-lang3:jar:3.7:compile
[INFO] +- org.apache.commons:commons-collections4:jar:4.1:compile
[INFO] +- org.slf4j:slf4j-api:jar:1.7.25:compile
[INFO] \- junit:junit:jar:4.11:test
[INFO]    \- org.hamcrest:hamcrest-core:jar:1.3:test
```

@ignazio1977 
I does not see any other dependency forcing an old httpclient version…
:/

I will investigate deeplier…

**Comment by sipi** [comment]
Excuse me for the disturbance, I just clean the affected dependencies from my m2 repository and It works :)

**Comment by ansell** [comment]
RDF4J seems to have issues internally with managing their Jackson versions in both of those dependency trees, as jackson-core should have the same version as jackson-databind, but that doesn't explain the HTTP Client classpath issue directly. It could be the cause if the 4 classloader stack frames include a jackson class. 

My PR to fix the RDF4J Jackson classpath issue using jackson-bom was merged fairly recently for RDF4J-2.3.x : https://github.com/eclipse/rdf4j/pull/818

OWLAPI manages the Jackson dependencies individually in the parent reactor dependencyManagement, but that may not be working as intended.
